tempesta_fw/http.c

```diff
@@ -2819,64 +3243,135 @@ tfw_http_req_process(TfwConn *conn, const TfwFsmData *data)
+	}
+
+	/* Forward message to destination. */
+	__HTTP_FSM_STATE(Http_Msg_Fwd) {
+		int err = 0;
 
 		/*
 		 * Add the request to the list of the client connection
 		 * to preserve the correct order of responses to requests.
 		 */
 		tfw_http_req_add_seq_queue(req);
 
+		if (unlikely(req->resp))
+			/* Response is already provided for the request. */
+			tfw_http_resp_fwd(req->resp);
+		else
+			err = tfw_cache_process(hmreq, tfw_http_req_cache_cb);
+
 		/*
 		 * The request should either be stored or released.
 		 * Otherwise we lose the reference to it and get a leak.
 		 */
-		if (tfw_cache_process((TfwHttpMsg *)req, tfw_http_req_cache_cb)) {
-			tfw_http_send_resp(req, 500, "request dropped:"
-					   " processing error");
+		if (err) {
 			TFW_INC_STAT_BH(clnt.msgs_otherr);
-			return TFW_PASS;
+			req->httperr.status = 500;
+			req->httperr.reason =
+				"request dropped: processing error";
+			__HTTP_FSM_JUMP(Http_Msg_Conn_Drop);
 		}
-
 		/*
-		 * According to RFC 7230 6.3.2, connection with a client
-		 * must be dropped after a response is sent to that client,
-		 * if the client sends "Connection: close" header field in
-		 * the request. Subsequent requests from the client coming
-		 * over the same connection are ignored.
-		 *
 		 * Note: This connection's @conn must not be dereferenced
-		 * from this point on.
+		 * from this point on if TFW_HTTP_F_CONN_CLOSE flag was set
+		 * and the message is not in streaming mode.
 		 */
-		if (req_conn_close)
-			break;
+		if (unlikely(stream_mode))
+			__HTTP_FSM_JUMP(Http_Msg_Stream);
+		__HTTP_FSM_EXIT_safe();
+	}
 
-		if (hmsib) {
-			/*
-			 * Switch connection to the new sibling message.
-			 * Data processing will continue with the new SKB.
-			 */
-			data_off = 0;
-			skb_len = skb->len;
-			conn->msg = (TfwMsg *)hmsib;
+	/*
+	 * Stream message parts to backend server or wait until
+	 * the request is fully received to send a response.
+	 */
+	__HTTP_FSM_STATE(Http_Msg_Stream) {
+		int err;
+
+		err = tfw_http_stream_req((TfwCliConn *)conn, req);
+		if (err) {
+			TFW_INC_STAT_BH(clnt.msgs_otherr);
+			req->httperr.status = 500;
+			req->httperr.reason =
+				"request dropped: streaming error";
+			__HTTP_FSM_JUMP(Http_Msg_Conn_Drop);
 		}
+		if (tfw_http_parse_stage(hmreq) != TFW_HTTP_PARSE_DONE)
+			__HTTP_FSM_EXIT();
+		__HTTP_FSM_JUMP(Http_Msg_End);
+	}
+
+	/*
+	 * Error during message processing, send response and close the
+	 * connection.
+	 */
+	__HTTP_FSM_STATE(Http_Msg_Conn_Drop) {
+		/*
+		 * There can be an existing response for the request, drop it
+		 * an error or security event occurred and send corresponding
+		 * error.
+		 */
+		tfw_http_msg_free(req->pair);
+		/*
+		 * Close connection immediately or after error response is sent.
+		 */
+		if (list_empty(&req->msg.seq_list))
+			r = tfw_client_drop(req, req->httperr.status,
+					    req->httperr.reason);
+		else
+			r = tfw_srv_client_drop(req, req->httperr.status,
+						req->httperr.reason);
```

Comment:
Initially, I mistakenly believed that `tfw_srv_client_drop()' will be used in server connections context only and gave it that awkward name. If we use this function in client connections context also, its name (in comparison with `tfw_slient_drop ()') does not reflect either its essence nor its purpose, is rather confusing and should be changed (maybe `tfw_client_drop_light` or something like that will be more appropriate).